### PR TITLE
Use post multi part form request for trees latest call

### PR DESCRIPTION
**Pull request by jatgarg**
1.) Start using the Multi Part form request(POST), so as to save the options call in odsp driver. Options call was fired whenever we had headers in the network request instead of query parameters. This happens when the length of request becomes greater than 2048. This was happening for about 80% of the calls.
2.) Use it only if enabled by use under a flag.

### CODE REVIEW

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (ram-ramasubramanian) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:363]: Nit: From a naming perspective, can we call this fetchSnapshotViaPost, instead of MultiPart, etc. which is too much implementation detail (that can change)?
Just a thought, definitely not blocking this PR either way.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (ram-ramasubramanian) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:542]: is this TODO still relevant?

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (ram-ramasubramanian) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:570]: Looks like some of Raj's w3c telemetry should be added here?

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (ram-ramasubramanian) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:589]: This is probably shareable code (with the old code paths).. If the idea is to remove the old code after veritying this, then this is probably ok.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (ram-ramasubramanian) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:371]: Double checking - this is called only for the GetLatest call I assume; none of the other calls to SPO (CreateNew, PostSummary,e tc.) use this POST API, correct?

Comment (ram-ramasubramanian) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:371]: To be clear, we can try to switch them over as well later if they are all hitting the OPTIONS calls; but we can use GetLatest to prove out that the approach works with a high degree of confidence.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (ram-ramasubramanian) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:308]: I assume this is the telemetry we will use to determine how effective the switch over is?

**Review by ram-ramasubramanian — CHANGES_REQUESTED** [review]

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:363]: I think fetchSnapshotViaPost is good.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:570]: I will merge this after his PR. So will pull his changes.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:371]: Yes, this is only for getLatest call. We can shift other calls to this form according to your plan.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:308]: This telemetry event if for refreshing the token if the original token is expired.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:308]: This is the one for post call failure and when callback code is called.:
this.logger.sendErrorEvent({ eventName: "TreeLatest_FallBackToNonMultiPartRequest" });

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:589]: Yes, we will remove the other code once we get confidence in POST call.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts]: Yes, this is not relevant now. Removed it.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (ram-ramasubramanian) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:373]: can we add one more query param (say _POst=1) to the body fo the POST request? Service can log that and use to quantify no OPTIONs from our telemetry.

**Review by ram-ramasubramanian — APPROVED** [review]

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:364]: "is" prefix means this function returns boolean, which is not the case.
Please rename to something that better reflects context

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:386]: You "forked" code based on POST vs. GET too early, which results in two instanced of code that deals with cache layer.
I believe prior workflow was better and we should revert, i.e. POST vs. GET should be encapsulated in fetchSnapshot() API - I do not see the reason to make that fork that early, and in getSnapshotTree()

**Review by vladsud — COMMENTED** [review]
Please take a look - I believe the old structure was better (in terms of flow and what we pivot first)

File: packages/drivers/odsp-driver/src/odspDocumentService.ts
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentService.ts]: Added to host storage policy.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts]: Added.

File: BREAKING.md
Comment (jatgarg) [reply BREAKING.md]: removed, I added so that while integrating they can enable it.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts]: Removed api.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts]: Changed the logic.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:471]: add error argument to sendErrorEvent()

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:467]: I'm not sure I agree with logic. Refresh happens if token has expired. If it's not then there is no second chance.
Given that we do not have fallback post to get on first attempt, I'm not sure why it makes sense to have fallback on second attempt. I.e. it seems like they need to be in sync - we either always do fallback, or we never do fallback.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:500]: Payload looks weird.
Why do we have guids around it?
Why do we need "post: 1"? I'm sure the fact that POST is used is available on the other side (though it might be a limitation of SPO where this info is lost in translation).

I'd inline crcl - code would be more readable.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:577]: Capitalization of property name was lost

**Review by vladsud — APPROVED** [review]

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:577]: Actually, it is consistent with the other property in this event, like all other properties are in small letters. So made consistent with other ones.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts]: A multi part form request looks like this where we use guids to make the form boundary. As confirmed with Ram, we can use _POST as they use _ to define internal vs external. They want to use this param for telemetry purpose so as to check whether they are getting any Options call around it.
Inlined crlf

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts]: Added

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts
Comment (jatgarg) [reply packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts]: So the logic now is we make a post call and if that fails for a reason for which getWithRetryForTokenRefresh will not retry we fallback to get call. otherwise we let getWithRetryForTokenRefresh to refresh token for post call and then we make post call again. Now second time if the post call fail again we fallback to get call.

**Review by gmRr7 — COMMENTED** [review]
R

**Pull request merged**
